controllers/apidefinition_controller_test.go

```diff
@@ -231,3 +241,145 @@ func TestEncodeIfNotBase64(t *testing.T) {
 		t.Fatalf("expect %s, got %s", in, s)
 	}
 }
+
+func TestProcessSubGraphExecution(t *testing.T) {
+	const (
+		subgraphName = "test-subgraph"
+		testSDL      = "sdl"
+		testSchema   = "schema"
+
+		newSubgraphName = "test-subgraph-new"
+		newTestSDL      = "new-sdl"
+		newTestSchema   = "new-schema"
+
+		testNs = "test-ns"
+	)
+
+	subGraph := &tykv1alpha1.SubGraph{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      subgraphName,
+			Namespace: testNs,
+		},
+		Spec: tykv1alpha1.SubGraphSpec{
+			SubGraphSpec: model.SubGraphSpec{
+				SDL:    testSDL,
+				Schema: testSchema,
+			},
+		},
+	}
+	newSubGraph := &tykv1alpha1.SubGraph{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      newSubgraphName,
+			Namespace: testNs,
+		},
+		Spec: tykv1alpha1.SubGraphSpec{
+			SubGraphSpec: model.SubGraphSpec{
+				SDL:    newTestSDL,
+				Schema: newTestSchema,
+			},
+		},
+	}
+
+	apiDef := tykv1alpha1.ApiDefinition{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      "test-name",
+			Namespace: testNs,
+		},
+		Spec: tykv1alpha1.APIDefinitionSpec{
+			APIDefinitionSpec: model.APIDefinitionSpec{
+				APIID: EncodeNS(types.NamespacedName{Name: "test-name", Namespace: testNs}.String()),
+				GraphQL: &model.GraphQLConfig{
+					GraphRef: subgraphName,
+				},
+			},
+		},
+	}
+
+	apiDefMalformed := &tykv1alpha1.ApiDefinition{}
+	apiDef.DeepCopyInto(apiDefMalformed)
+	apiDefMalformed.ObjectMeta.Name = "test-api-malformed"
+	apiDefMalformed.Spec.GraphQL = nil
+
+	apiDefLinkedMultiple := &tykv1alpha1.ApiDefinition{}
+	apiDef.DeepCopyInto(apiDefLinkedMultiple)
+	apiDefLinkedMultiple.ObjectMeta.Name = "test-name-multiple-link"
+	apiDefLinkedMultiple.Spec.APIID = EncodeNS(client.ObjectKeyFromObject(apiDefLinkedMultiple).String())
+
+	apiDefNewGraphRef := &tykv1alpha1.ApiDefinition{}
+	apiDef.DeepCopyInto(apiDefNewGraphRef)
+	apiDefNewGraphRef.Spec.GraphQL.GraphRef = newSubgraphName
+
+	objects := []runtime.Object{&apiDef, apiDefMalformed, apiDefLinkedMultiple, subGraph, newSubGraph}
+	eval := is.New(t)
+
+	cl, err := NewFakeClient(objects)
+	eval.NoErr(err)
+
+	testCases := []struct {
+		testName    string
+		apiDef      *tykv1alpha1.ApiDefinition
+		subGraph    *tykv1alpha1.SubGraph
+		apiMutator  func(definition *tykv1alpha1.ApiDefinition)
+		expectedErr error
+	}{
+		{
+			testName: "processing malformed ApiDefinition with nil GraphQLConfig field",
```

Comment:
Having empty GraphQLConfig is valid case. Should we remove word malformed?

Reply:
I meant nil GraphQLConfig is malformed for `processSubGraphExec` method.

Reply:
okay. It's good then